### DISCUSSION TITLE
Add customized new features

### DISCUSSION THREAD

**Discussion by FateDecider** [qcha41/usit#32] [Q&A] [answered]
Hi there, I would like to add some features customized to my purpose. But I will need some guidance on which parts of the code to change. Let me explain the two things I have in mind:

1. For a multi-parameter scan, I want to reverse the order of the list corresponding to the final parameter, for every other value of the second last parameter. To do this, I have this in scan.py > execRecipe:
```
        # iter over each parameter (do once if no parameter!)
        for i, paramValueList in enumerate(product(*paramValues_list)):
            if self.reverse_order:  # Check if reverse order is enabled
            # Access the value of the second last parameter
                if len(paramValues_list) >= 2 and len(paramValues_list[-1]) > 1 and len(paramValues_list[-2]) > 1:
                    second_last_parameter_value = paramValues_list[-2][i]  
                    if second_last_parameter_value % 2 != 0:  # Check if the value is odd
                        paramValueList[-1] = reversed(paramValueList[-1])  # Reverse the order of the list for the last parameter
```
I want to add a button to the scanning GUI that sets self.reverse_order = True. Can you advise how to add this button in the respective backend codes?

2. For a multi-parameter scan, I have two parameters to scan and I measure a variable, and I want that to be plotted in "Scan" figure as a 2D plot instead of 1D line plots.  I believe I have to change this setting in gui\scanning\figure.py, but not entirely sure how.  Please advise. 

Thanks for the advice!

**Comment by Python-simulation** [comment]
Hi!

1. I am not sure what you meant by "to reverse the order of the list corresponding to the final parameter, for every other value of the second last parameter". **Could you provide an example** with values so I can see if it is possible to do what you want with the current Autolab version (using $eval: and Variables)?
To add a button you can either use QtDesigner to open the interface.ui file, add a button and connect to it in any scanning file. Or simpler, using code:
```
# In scanning/main.py at the end of __init__
self.reverseCheckBox = QtWidgets.QCheckBox("Reverse order")
self.reverseCheckBox .stateChanged.connect(self.doSomething)
self.verticalLayout_recipe.addWidget(self.reverseCheckBox)

def doSomething(self):
    self.reverse_order = self.reverseCheckBox.isChecked()  # boolean to use for scan
   # You can then use this state in scan.py BUT you need to add it to ScanThread that doesn't communicate with the GUI (as an argument maybe)
```

2. It's something I'd also like to do but have never found the time.
Currently, plotting is done by selecting the x and y axis:
```
x = subdata.loc[:,variable_x]
y = subdata.loc[:,variable_y]
```
and plotting a line:  `curve = self.ax.plot(x, y, ...)`.
We should add a way to select x,y,z and plot it as an image like for true images: `self.figMap.setImage(image_data, ...)`
I didn't plan to do it soon so I can't help you more on this topic sorry.

**Comment by FateDecider** [comment]
Hi, thanks for the prompt reply. I am planning to scan the positions like this (in black) instead of (in green), which is the default for the nested loop structure
![image](https://github.com/autolab-project/autolab/assets/57542293/5dd2a82d-2daa-4f0e-9909-91b5894244e0)

**Reply by Python-simulation** [reply]
Ok I see.
I'm sure you can find a way to do it based on the following example:
![image](https://github.com/autolab-project/autolab/assets/46069237/d05eb29d-814c-43a9-a46c-e031e8d6890d)
I will see tomorrow for an exact solution, to tired for tonight ^^

**Reply by Python-simulation** [reply, marked as answer]
Solution that works only for linear step and by providing the maximum value of the x parameter inside the recipe: 
![image](https://github.com/autolab-project/autolab/assets/46069237/e4fa2ef4-ee87-4cd0-9d63-a2a2f90d5687)

**Reply by Python-simulation** [reply]
For previous Autolab versions without Variable, you can do the same by replacing the parameter name with its address (dummy.constant(), system.parameter_buffer() ...)

**Reply by FateDecider** [reply]
Thanks!

For my second request, I cheated by doing the following:
![image](https://github.com/autolab-project/autolab/assets/57542293/58309a3f-17ed-4985-a8a6-20f275f59305)
and I save the values I need in the Driver temporarily and assemble them using a method in Driver():
![image](https://github.com/autolab-project/autolab/assets/57542293/9abf1e2a-c286-4973-a6ce-f62c985f5981)
but I believe the image did not quite plot correctly (I think my array formatting/shape is not correct - for my purpose, 2D image is good)
![image](https://github.com/autolab-project/autolab/assets/57542293/431fbc34-e44b-4f66-ba24-c8910f26ef7a)

**Reply by Python-simulation** [reply]
The issue comes from the fact that you tried to plot three 'images' with different shapes.
The variable data correspond to a list with an image per scan step (in your example 3). The first image you try to plot is array([[0.01503037]]) with shape (1,1) (so not an image). The second is None and the third is array([[0.01503037, 0.02105777, 0.01424197]]) with shape (1,3).
The way images are plotted require to have an image per step but you tried to create an image for a whole scan if I understood correctly.

**Reply by Python-simulation** [reply]
One solution would be to create an empty array with the correct size and populate each point during the scan to have something like that:
```
data = [array([[0.01503037, None, None]]), array([[0.01503037, 0.02105777, None]]), array([[0.01503037, 0.02105777, 0.01424197]])]
```
I remind that `data=[image_1, image_2, image_3]` for a scan with 3 steps.

**Reply by Python-simulation** [reply]
I tried to do 2D plot for scan (not commited yet). Here is the first result:
![image](https://github.com/autolab-project/autolab/assets/46069237/c4955878-6463-42bc-9174-bfc521d87c53)
I'm not totally happy with it because it is ordered by step ID (from bottom-left to top-right) and doesn't take into account the x and y values. I saw there was a [pcolormesh](https://pyqtgraph.readthedocs.io/en/latest/api_reference/graphicsItems/pcolormeshitem.html) equivalent in pyqtgraph but didn't try yet, I expect to have the x and y scale with it. Thoughts?

**Reply by FateDecider** [reply]
I honestly think this is good enough for my purpose. Maybe leaving this as a grid based on the combination of step ID is sufficient because there could be scenarios we do custom array and it wouldn't make sense to label everything in chronological order. I also like we can change the X, X2, and Y axes. This capability is very flexible. Could you push this to a separate branch for me to test for my experiments? Thanks a lot for your prompt help!

**Reply by Python-simulation** [reply]
You can test it in branch [scan_2D_image](https://github.com/autolab-project/autolab/tree/scan_2D_image). Currently the data plotted is only 'correct' when you select x, y, z, but gives a wrong sorting if you plot y, x, z.
Edit: should be good now (except if select log scale)

**Reply by FateDecider** [reply]
By the way, this turns out pretty good. Thank you!! I am wondering if it is conceptually possible to have Y axis as an array so we can tune which value of the Y-axis array we want to display as the z-value of the colormesh?

**Reply by Python-simulation** [reply]
It is possible and I though about it already.
The idea would be to add a filter option to filter all values which have the same y_value in the scan dataframe and to plot as a line two other variables (x_variable, z_variable).
But I don't know if and when I will do it.

You can look in the mean time at [pandasgui](https://pypi.org/project/pandasgui/) which propose a GUI with filtering and plotting

**Reply by Python-simulation** [reply]
@FateDecider, I've added in 6303fc3 a filter option to the scanner. I'll add a slider later to set the value in a more user-friendly way than QLineEdit.

**Reply by FateDecider** [reply]
great cool, i just checked it out. for the Z-axis, is it possible to select a np.ndarray type object? i want to be able to select the particular element from this np.ndarray as the value for the Z-axis over the grid of (X,Y) coordinates.
![image](https://github.com/user-attachments/assets/fa7d3f7f-500c-4d76-8ca1-1f94dfe610f8)

**Reply by Python-simulation** [reply]
Hi! The '2D plot' option is linked to the scan dataframe which is a collection of int or float element for each step (each id). If you measure any value for a (x,y,z) coordinate, you can plot the variable as an image versus (x,y) and filter z (z == value). But if your measurement is an array, you cannot use '2D plot' but you can use the filter to select which line to plot.
